extensions/csv/src/main/java/io/deephaven/csv/reading/CellGrabber.java

```diff
@@ -0,0 +1,276 @@
+                // "", to be interpreted as ". So we'll spill this string (up to the first quotation mark),
+                // skip the second quotation mark, and keep going.
+                spillRange();
+                // Skip the second quotation mark.
+                ++offset;
+                startOffset = offset;
+            }
+            if (!tryEnsureMore()) {
+                throw new CsvReaderException("Cell did not have closing quote character");
+            }
+        }
+        // We got out of the quoted string. Remember where you are and then consume the
+        // rest of the characters.
+        lastInRow[0] = finishField(dest);
+
+        // The easiest way to make all the above logic run smoothly is to let the final quotation mark
+        // (which will unconditionally be there) and subsequent whitespace (if any) into the field.
+        // Then we can simply trim it back out now.
+        while (dest.begin() != dest.end() && RangeTests.isWhitespace(dest.back())) {
+            dest.setEnd(dest.end() - 1);
+        }
+        if (dest.begin() == dest.end() || dest.back() != quoteChar) {
+            throw new RuntimeException("Logic error: final non-whitespace in field is not quotation mark");
+        }
+        dest.setEnd(dest.end() - 1);
+    }
+
+    private void processUnquotedMode(final CharSlice dest, boolean[] lastInRow) throws CsvReaderException {
+        startOffset = offset;
+        lastInRow[0] = finishField(dest);
+    }
+
+    private void skipWhitespace() throws CsvReaderException {
+        do {
+            while (offset != size) {
+                final char ch = buffer[offset];
+                if (!RangeTests.isWhitespace(ch)) {
+                    return;
+                }
+                ++offset;
+            }
+        } while (tryEnsureMore());
+    }
+
+    /**
+     * @return true iff the field is the last in the row, false otherwise.
+     */
+    private boolean finishField(final CharSlice dest) throws CsvReaderException {
+        while (true) {
+            while (offset != size) {
+                final char ch = buffer[offset];
+                if (ch == fieldDelimiter) {
+                    finish(dest);
+                    ++offset;
+                    return false;
+                }
+                if (ch == '\n') {
+                    finish(dest);
+                    ++offset;
+                    return true;
+                }
+                if (ch == '\r') {
+                    finish(dest);
+                    ++offset;
+                    if (tryEnsureMore()) {
+                        // might be \r\n
+                        if (buffer[offset] == '\n') {
+                            ++offset;
+                        }
+                    }
+                    return true;
+                }
+                ++offset;
+            }
+            if (!tryEnsureMore()) {
```

Comment:
Similar comment as above for two nested whiles versus if condition.

Reply:
Good catch, thanks. Changed.